src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1744,9 +1758,25 @@ protected void AfterMembersChecks(BindingDiagnosticBag diagnostics)
                 }
             }
 
-            if (IsFileLocal && (object?)ContainingType != null)
+            if (IsFileLocal)
             {
-                diagnostics.Add(ErrorCode.ERR_FileTypeNested, location, this);
+                // A well-behaved file-local type only has declarations in one syntax tree.
+                // There may be multiple syntax trees across declarations in error scenarios,
+                // but we're not interested in handling that for the purposes of producing this error.
+                var tree = declaration.Declarations[0].SyntaxReference.SyntaxTree;
+                // TODO2: we probably don't want to linearly search the syntax trees.
```

Comment:
I looked a bit at caching the set of "used file paths" on `SyntaxAndDeclarationManager`, but since that type is designed to incrementally update its collections when syntax trees are added/removed/replaced, it wasn't obvious to me what representation should be used for that info.

I think in order to be incremental there, we would need to track a dictionary of "file path" to "how many times the file path is used", in order to e.g. implement "remove" correctly. Otherwise we might need to generate the set from scratch in certain scenarios. See https://github.com/dotnet/roslyn/commit/0fa4408b514be1d8a8c0bbb2d5451087499f150a